% !TEX root = thesis.tex
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Materials and Methods
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\chapter{Material and Methods}
  \label{sec:matmet}

\section{Implementation of FracMinHash}
The calculation of phylogenetic outlines using FracMinHash was implemented based
on the libraries \texttt{java17-openjdk} \cite{theopenjdkcommunityJDK172021},
\texttt{jloda3} 1.0.0, \texttt{splitstree6} 1.0.0
\cite{husonApplicationPhylogeneticNetworks2006} and
\texttt{openapi-generator-maven-plugin} 6.3.0
\cite{openapi-generatorcontributorsOpenAPIGenerator2023}. Note that I will
denote the library as \texttt{splitstree6} and the tool that is mainly based on
that library as \texttt{SplitsTree}.

To sketch a sequence, its $k$-mers are read, $k$-mers that include ambiguous
bases (i.e. bases that are "N" or "n") are skipped. FASTA files can contain
multiple sequences, but they are not concatenated in \texttt{fmhdist} and the
$k$-mer decomposition is done for each sequence individually, i.e. no $k$-mer
spans two or more sequences. The $k$-mers are brought into their canonical form
by transforming all characters to uppercase, calculating the reverse complement
of the $k$-mer and using the alphabetically smaller value of both
\cite{ondovMashFastGenome2016,irberLightweightCompositionalAnalysis2022} .

The $k$-mers are then hashed using a given hash function and a random seed. For
this, the \texttt{LongHashFunctions} from \texttt{zero-allocation-hashing} 0.16
are used \cite{ZeroAllocationHashing2022}. Java \texttt{long} values are signed,
thus the range of the hash function is $[-2^{63}, 2^{63}]$. Therefore, the
threshold can be calculated using $\frac{2^{64}}{s} -2^{63}$. 

Hash values that are smaller than that threshold are kept in the sketch. The
sketch is saved in a serialized format on the hard disk. To estimate the Jaccard
index, the containment index and from those the distances, the formulas
discussed in Chapter~\ref{sec:background} were used
\cite{heraDerivingConfidenceIntervals2023,irberLightweightCompositionalAnalysis2022}.

The distances are stored in Nexus file format using the
\texttt{splitstree6} 1.0.0 \cite{husonApplicationPhylogeneticNetworks2006}
serialization capabilities.

To store the resulting outlines as SVG images, \texttt{jfree.svg} 5.0.5
\cite{gilbertJFreeSVG2023} was used.


\section{Benchmarking hash functions}
To get an understanding of which hash function performs best in terms of
runtime, the \texttt{jmh} benchmarking framework 1.37
\cite{theopenjdkcommunityJavaMicrobenchmarkHarness2023} was used to benchmark
the following hash functions initially described by the \texttt{smhasher}
library (MurMur3) \cite{applebyAapplebySmhasher}, by the \texttt{xxHash} library
(XXH3, XXH64, XXH128) \cite{colletXxHash2023}, by the \texttt{cityhash} library
(cityhash 1.1) \cite{pikeCityhash2011}, by the \texttt{farmhash} library
(farmhash 1.0, farmhash 1.1) \cite{pikeFarmhash2014}, by the \texttt{wyhash}
library (wyhash 3) \cite{wangWyhash2019} and by the \texttt{MetroHash} library
\cite{rogersMetroHash2018}.
 
While other hash function implementations were considered, the implementation of
all hash functions is given by \texttt{zero-allocation-hashing} 0.16
\cite{ZeroAllocationHashing2022}. The benchmark was executed in throughput mode
using a fork count of 2. Besides that, default parameters were used. Each
benchmark consisted of hashing a string of length 21, which is a $k$-mer size
used in different applications
\cite{ondovMashFastGenome2016,bagciMicrobialPhylogeneticContext2021}. The
benchmark was executed on a consumer notebook with 16GB memory and 8 cores
($4.2$ GHz).


\section{Datasets}
To explore properties of FracMinHash, analysis was performed using different
datasets. 

The first dataset (\textbf{A}) consists of 128 different \textit{Phytophthora}
genomes. The list is taken from \cite{mandalComparativeGenomeAnalysis2022}
without further modifications. Genomes were downloaded from NCBI using the
\texttt{datasets} utility \cite{sayersDatabaseResourcesNational2022} which was
given the accession codes listed in that study.

The second dataset (\textbf{B}) consists of 72 mtDNA sequences of
\textit{Phytophthora}
and other Oomycetes of the Peronosporaceae family. The list is taken without
further modifications from Supplementary Table 1 from
\cite{winkworthComparativeAnalysesComplete2022}. Genomes were downloaded from
NCBI with the Entrez interface \cite{sayersDatabaseResourcesNational2022} using
the accession codes listed in that study, appended by the identifier of the most
recent version for that accession code.

The third dataset (\textbf{C}) consists of all 64 \textit{Phytophthora}
reference sequences in the NCBI databasase as well as five different query
sequences that are typically found in soil samples of avocado orchards that are
infected with \textit{Phytophthora cinnamomi}
\cite{solis-garciaPhytophthoraRootRot2020}. Those query sequences are divided
into two bacterial genomes, two fungal and one \textit{Phytophthora cinnamomi}
genome:

\begin{itemize}
  \item \textit{Mortierella claussenii} (GCA\_022750515.1), fungal, $37.5$ Mb
  \item \textit{Phytophthora cinnamomi} (GCA\_001314365.1), $54.7$ Mb  
  \item \textit{Pseudomonas syringae} (GCA\_018394375.1), bacterial, $5.9$ Mb
  \item \textit{Thermogemmatispora aurantia} (GCA\_008974285.1), bacterial, $5.6$ Mb
  \item \textit{Venturia carpophila} (GCA\_014858625.1), fungal, $35.9$ Mb
\end{itemize}

The full list of reference sequences can be found in the appendix
\ref{sec:refseq}. The reference sequences were downloaded using the NCBI web
interface (Filter: reference sequences), the query sequences were downloaded
using the \texttt{datasets} \cite{sayersDatabaseResourcesNational2022} utility.



\section{Comparison with published phylogenies}
Figure 4 in \cite{mandalComparativeGenomeAnalysis2022} displays a rooted
phylogenetic tree that was generated using \texttt{mashtree}
\cite{katzMashtreeRapidComparison2019,ondovMashFastGenome2016}. Unfortunately,
the corresponding distance matrix or a serialized version of the tree are not
available. Thus, the data needs to be re-computed. For this, the
\texttt{mashtree\_bootstrap.pl} 1.4.6 was applied to dataset \textbf{A} as
described by
\cite{mandalComparativeGenomeAnalysis2022}, additionally the distance matrix was
saved using the \texttt{--outmatrix} parameter.

Distance matrices for dataset \textbf{A} using the FracMinHash method
\cite{heraDerivingConfidenceIntervals2023,irberLightweightCompositionalAnalysis2022}
were calculated with \texttt{fmhdist} using different combinations of the
scaling parameter $s$ and $k$-mer size $k$. For a fixed $k=21$, different $s \in
\{500, 1000, 2000, 4000\}$ were applied, for a fixed $s=2000$, different $k \in
\{10, 19, 20, 21, 25, 30\}$ were applied.

To ensure comparability with the \texttt{mashtree} results, the same hashing
function (MurMur3 \cite{applebyAapplebySmhasher,ZeroAllocationHashing2022}) and
random seed (42) were used to calculate the sketches
\cite{katzMashtreeRapidComparison2019,ondovMashFastGenome2016}.

For all distance matrices, \texttt{SplitsTree} 6.3.20
\cite{husonApplicationPhylogeneticNetworks2006} was used to obtain trees, splits
and outlines. Trees were calculated using the Neighbor Joining method
\cite{saitouNeighborjoiningMethodNew1987}. Splits were obtained using the
Neighbor Net method
\cite{bryantNeighborNetAgglomerativeMethod2004,bryantNeighborNetImprovedAlgorithms2023}
using the default parameters. Based on this, a phylogenetic outline was
calculated \cite{bagciMicrobialPhylogeneticContext2021}. Trees were compared
using  the Consensus Network method \cite{hollandUsingConsensusNetworks2004}
with default options, except \texttt{Consensus=ConsensusNetwork}.



\section{Analysis of split differences}
\label{sec:splitanalysis}
To further explore the differences between the phylogenetic outline based on the
Mash distances and those based on FracMinHash, the splits were analysed in
detail. For this, the splits were exported from \texttt{SplitsTree} 6.3.20
\cite{husonApplicationPhylogeneticNetworks2006} using the plain text format.

Those files were processed with the script \texttt{compare\_splits.py}. This is
based on \texttt{python} 3.12 \cite{vanrossumPythonReferenceManual2009} and
\texttt{pandas} 2.1.4
\cite{PandasdevPandasPandas2024,mckinneyDataStructuresStatistical2010}. All sets
of splits are compared pairwise, that is for two sets $\Sigma_A$ and $\Sigma_B$,
the following properties are calculated:

\begin{itemize}
  \item the total sum of weight for all splits in $\Sigma_A$ and $\Sigma_B$,
  respectively: $\sum_{s \in \Sigma_A}{\omega(s)}$ and $\sum_{s \in
  \Sigma_B}{\omega(s)}$ 
  \item the number of splits in $\Sigma_A / \Sigma_B$ and the number of splits
  in $\Sigma_B / \Sigma_A$
  \item the total weight of the split differences, i.e.
  $\sum_{s \in \Sigma_A / \Sigma_B}{\omega(s)}$ and $\sum_{s \in \Sigma_B /
  \Sigma_A}{\omega(s)}$
  \item the Robinson-Foulds distance
  \cite{robinsonComparisonPhylogeneticTrees1981} of the two sets of splits, i.e.
  $D_{RF} = \frac{1}{2}|(\Sigma_A / \Sigma_B) \cup (\Sigma_B / \Sigma_A)|$
\end{itemize}

The script outputs a list of the diverging splits sorted by the weight of the
split. Each split is formatted as a list of taxon names on the smaller side of
the split sorted alphabetically and joined using the pipe symbol ("|") such that
the split can be searched in \texttt{SplitsTree} 6.3.20 using the regular
expression search.



\section{Analysis for shorter sequences}
To get an idea of the practical lower boundaries of FracMinHash in terms of
input genome size, dataset \textbf{B} was sketched. As this dataset consists of
just a single long FASTA file, it was split into its records using
\texttt{split-fasta} 1.0.0 \cite{vashishtSplitfasta2020}. The files were
sketched using FracMinHash
\cite{irberLightweightCompositionalAnalysis2022,heraDerivingConfidenceIntervals2023}
in \texttt{fmhdist} using $k=21$ and the FarmHash 1.0 hash function
\cite{pikeFarmhash2014,ZeroAllocationHashing2022} with random seed 42. The
scaling parameter was selected with  $s \in \{1, 10, 50, 100, 1000\}$ such that
a comparison of the results for different scaling parameters is possible.

Given the calculated distances, \texttt{SplitsTree} 6.3.20
\cite{husonApplicationPhylogeneticNetworks2006} was used to obtain trees, splits
and outlines. Trees were calculated using the Neighbor Joining method
\cite{saitouNeighborjoiningMethodNew1987}. Splits were obtained using the
Neighbor Net method
\cite{bryantNeighborNetAgglomerativeMethod2004,bryantNeighborNetImprovedAlgorithms2023}
using the default parameters. Based on this, a phylogenetic outline was
calculated \cite{bagciMicrobialPhylogeneticContext2021}.


\section{Phylogenetic outlines for distantly related species}
To explore the claim that FracMinHash works better for genomes of different
sizes, both reference and query sequences of dataset \textbf{C} were sketched,
distances
calculated and the results compared. 

\subsection*{Sketching using FracMinHash}
\label{sec:intersectionlog}
Reference and query sequences were sketched using \texttt{fmhdist} with $k=21$,
$s=2000$ using FarmHash \cite{pikeFarmhash2014,ZeroAllocationHashing2022} with
random seeds $rs \in \{10, 20, 30, 40, 50\}$. As a Mash sketch size
$s_{mash}=10000$ is typically found in literature, e.g. as the default value
used in \texttt{mashtree}
\cite{katzMashtreeRapidComparison2019,ondovMashFastGenome2016} and to ensure
that it is not the fact that the relevant sketches are just larger, additional
sketching with $s=500$ (which aims to bring the sketch size of the bacterial
query sequences to 10000) and $s=3500$ (which aims to bring the sketch size of
the fungal query sequences to 10000) was performed. Using the sketches, the
distance matrix was calculated. For this, the implementation was changed such
that an additional log is created that lists all empty intersections when
calculating the intersection of the two sketches.

\subsection*{Sketching using MashTree}
To obtain a base value to compare against, \texttt{mashtree} 1.4.6
\cite{katzMashtreeRapidComparison2019,ondovMashFastGenome2016} was applied with
hash seed $rs \in \{10, 20, 30, 40, 50\}$, \texttt{--outmatrix} and the
\texttt{--save-sketches} parameter. The resulting sketches were then also
converted into JSON format using \texttt{mash info -d}
\cite{ondovMashFastGenome2016}.

\subsection*{Visual comparison and split differences}
Given all calculated distances, \texttt{SplitsTree} 6.3.20
\cite{husonApplicationPhylogeneticNetworks2006} was used to obtain trees, splits
and outlines. Trees were calculated using the Neighbor Joining method
\cite{saitouNeighborjoiningMethodNew1987}. Splits were obtained using the
Neighbor Net method
\cite{bryantNeighborNetAgglomerativeMethod2004,bryantNeighborNetImprovedAlgorithms2023}
using the default parameters. Based on this, a phylogenetic outline was
calculated \cite{bagciMicrobialPhylogeneticContext2021}.

The splits were then also analysed using the script outlined in Section
\ref{sec:splitanalysis}.

\subsection*{Obtaining reference values}
As there are no published phylogenies for dataset \textbf{C} that enable a
comparison, I
need a different ground truth to compare the results of above distance
calculation against. As the aim for this experiment is to analyse the influcence
of different genome sizes of distantly related organisms, I have limited this
analysis to the five query sequences of dataset \textbf{C} and the reference
genome for
\textit{Phytophthora infestans} (GCF\_000142945.1) as this is one of the largest
genomes in the set of reference sequences. For those sequences, I have prepared
two different values to compare against.

The first is the ANI calculated with \texttt{OrthoANI}
\cite{leeOrthoANIImprovedAlgorithm2016} using the default parameters with
\texttt{BLAST+} 2.14.1 \cite{camachoBLASTArchitectureApplications2009} for
alignment.

The second is the Average Amino Acid Identity (AAI)
\cite{konstantinidisGenomeBasedTaxonomyProkaryotes2005} calculated with the
\texttt{Enveomics Collection} online resource
\cite{rodriguez-rEnveomicsCollectionToolbox2016} using default parameters. As
this requires amino acid sequences as input and those are not available for the
\textit{Venturia carpophila} and \textit{Phytophthora cinnamomi} sequences, they
were predicted using the \texttt{gmes\_petap.pl} script provided in the
\texttt{GeneMark ES} 4.72 package \cite{lomsadzeGeneIdentificationNovel2005}
with the default parameters and extracted using the \texttt{gffread} utility
0.12.7 \cite{perteaGFFUtilitiesGffRead2020}.


\subsection*{Calculating empty intersections}
When the sketch intersections are empty, the estimated Jaccard similarity is
always 0. However, Mash includes not only the sketch intersections in the
Jaccard estimation, but also the sketch of the union of the input genomes, i.e.
$|S(A \cup B) \cap S(A) \cap S(B)|$ \cite{ondovMashFastGenome2016}. This enables
Jaccard estimations of 0 even though the sketch intersections are not empty.

To explore if FracMinHash produces more or less empty intersections, I have
prepared the script \texttt{get\_empty\_intersections.py} which takes a list of
Mash sketches in JSON format and outputs all pairs for which the numerator of
the Jaccard estimation is $0$. One could have also taken all entries in the
distance matrix that are $1$, but to ensure that no pruning or rounding is
happening at any place, I have calculated the numerator directly.

The same is possible using the added log output generated by the distance
calculation described in Section \ref{sec:intersectionlog}.

Using five different random seeds, this output can be converted into a table
that lists the number of empty intersections for each pairwise calculation.


\section{Hash Count analysis}
\subsection*{$k$-mer Coordinates}
To analyse the distribution of $k$-mers that are part of a FracMinHash sketch,
the coordinates of each $k$-mer were exported when sketching dataset \textbf{C}
in \texttt{fmhdist} with $k=21$, $s=2000$ and FarmHash 1.0 with random seed $rs
\in \{10, 20, 30, 40, 50, 60, 70, 80, 90\}$. The larger amount of different hash
seeds in comparison with the other experiments further reduces the impact of a
single seed. The coordinates saved for each $k$-mer consist of the following:

\begin{itemize}
  \item the sequence index, the index of the sequence inside the FASTA file
  \item the $k$-mer index relative to the file, $k$-mers including
  ambiguous bases are not counted, no $k$-mer is spanning two or more sequences
  in the FASTA file
  \item the $k$-mer index relative to the current sequence, $k$-mers including
  ambiguous bases are not counted
  \item the $k$-mer index relative to the file, $k$-mers including ambiguous
  bases are counted
  \item the $k$-mer index relative to the current sequence, $k$-mers including
  ambiguous bases are counted
  \item the $k$-mer itself, not the canonical form of it
  \item the hash value of the canonical $k$-mer
\end{itemize}

The coordinate files were analysed with the script
\texttt{coordinates\_complexity\_correlation.py} using \texttt{python} 3.12
\cite{vanrossumPythonReferenceManual2009} and \texttt{numpy} 1.26.3
\cite{ArrayProgrammingNumPy}. For this analysis, each sequence in the FASTA file
is split into non-overlapping windows of size $w=10000$. For each window, the
number of hashes inside the window that are part of the sketch are counted
including duplicates and ambiguous bases.

\subsection*{Sequence Complexity}
Also part of the analysis is the sequence complexity for each window. For this,
the \texttt{macle} tool \cite{pirogovHighcomplexityRegionsMammalian2019} was
used to obtain the sequence complexity $C_m$. First, the single FASTA file of
each genome was split into the individual sequences using \texttt{split-fasta}
1.0.0 \cite{vashishtSplitfasta2020}. Then, the complexity was calculated using
the default parameters of \texttt{macle} and a window size of $w=10000$. The
output identifies each window by the position of its center base, so this needs
to be shifted to the front of the window to match it to the corresponding hash
count window. \texttt{macle} outputs a complexity of $-1$ if the window contains
ambiguous bases, this information was used to identify and discard those
windows.

\subsection*{Statistical Analysis of connections between hash counts and sequence complexity}
For all windows with a non-negative sequence complexity $C_m$, the Pearson
Correlation $r$ was calculated to evaluate a potential correlation between hash
counts and sequence complexity.

Next, to check if there is any connection between windows having an unusual hash
count and having a low sequence complexity, we need to define what is considered
unusual: Given the scaling parameter $s$ and a window size $w$, we would expect
on average $\frac{w}{s}$ hashes per window that are part of the sketch. A window
has \textit{unusual hash counts} if the number of hashes is outside the interval
$[\frac{w}{s} - r \frac{w}{s}, \frac{w}{s} + r \frac{w}{s}]$. I have set
$r=0.95$, $s=2000$ and $w=10000$, which classifies all windows with a hash count
of $0$ or $\geq 10$ as unusual.

Given those inputs, I have performed the Mann-Whitney U test
\cite{mannTestWhetherOne1947,wilcoxonIndividualComparisonsRanking1945} with the
complexity values of each window split into the two categories "unusual count"
and "usual count". $\alpha$ was corrected with $\alpha = \frac{0.05}{69}$ to
account for multiple testing of the 69 different genomes.

The statistics were computed using \texttt{python} 3.12
\cite{vanrossumPythonReferenceManual2009} and \texttt{scipy} 1.11.3
\cite{virtanenSciPyFundamentalAlgorithms2020}.

\subsection*{Analysis of potential connection between low hash counts and gene location}
To check if unusual hash counts can be linked to the location of (effector)
genes in \textit{Phytophthora infestans}, I have mapped the start position of
each coding sequence (CDS) annotation to the corresponding window and counted
the number of CDS per window.

\section{Runtime benchmarks}
To compare the runtime of \texttt{fmhdist} with \texttt{mash} and
\texttt{sourmash}, each tool was executed five times to calculate the sketches
for dataset \textbf{C} on a consumer notebook using 16GB memory and up to 8
cores ($4.2$ GHz).

The performance benchmark is limited to the creation of the sequence sketches
as this is by far the most time consuming part of the pipeline. The runtime was
measured using the wall clock output of the \texttt{time} utility shipped with
most Linux distributions. Unless otherwise specified, default settings were
used. 

The runtime for \texttt{mash} was measured using \texttt{time mash sketch -l
input.csv -s 10000} for the single-threaded use case and \texttt{time mash
sketch -l input.csv -s 10000 -p 6} for the multithreaded use case. The runtime
for \texttt{sourmash} was measured using \texttt{time sourmash sketch dna
--from-file input.csv}. The runtime for \texttt{fmhdist} was measured using
\texttt{time java -jar fmhdist.jar --input input.csv -t 1 -k 31 -s 1000} for the
single-threaded use case and \texttt{time java -jar fmhdist.jar --input
input.csv -t 6 -k 31 -s 1000} for the multithreaded use case. \texttt{sourmash}
currently only supports single threaded execution out of the box, so only this
scenario could be benchmarked.
