% !TEX root = thesis.tex
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Materials and Methods
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\chapter{Material and Methods}
  \label{sec:matmet}

\section{Implementation of FracMinHash}
The generation of phylogenetic outlines using FracMinHash was implemented using
\texttt{java-17}, \texttt{jloda3} 1.0.0, \texttt{splitstree6} 1.0.0
\cite{husonApplicationPhylogeneticNetworks2006} and
\texttt{openapi-generator-maven-plugin} 6.3.0.

To sketch a sequence, its $k$-mers are read. $k$-mers that include ambiguous
bases (that is, bases that are "N" or "n") are skipped. FASTA files can contain
multiple sequences, but they are not concatenated and the $k$-mer decomposition
is done for each sequence individually, i.e. no $k$-mer spans two or more
sequences. The $k$-mers are brought into their canonical form
\cite{ondovMashFastGenome2016,irberLightweightCompositionalAnalysis2022} by
transorming all characters to uppercase, calculating the reverse complement of
the $k$-mer and using the alphabetically smaller value of both.

The $k$-mers are then hashed using a given hash function and a random seed. For
this, the \texttt{LongHashFunctions} from \texttt{zero-allocation-hashing} 0.16
are used. \todo{Check: Do I need a full list or just the main contributers? Do I
need a reference for each?} Java \texttt{long} values are signed, so the range
of the hash function is $[-2^{63}, 2^{63}]$. Thus, the threshold can be
calculated using $-2^{63} + \frac{2^{64}}{s}$. 

Hash values that are lesser than that threshold are kept in the sketch. The
sketch is saved in a serialized format on the hard disk. To estimate the Jaccard
index, the containment index and from those the distances, the formulas
discussed in section \ref{sec:formulas} were used
\cite{heraDebiasingFracMinHashDeriving2023,irberLightweightCompositionalAnalysis2022}.

The distances are then stored in Nexus file format using the \texttt{SplitsTree}
6.0.0-alpha \cite{husonApplicationPhylogeneticNetworks2006} serialization
capabilities.


\section{Benchmarking the Hash Functions}
To get an understanding of which hash function performs best in terms of
runtime, the \texttt{jmh} benchmarking framework 1.37 was used to benchmark the
following hash functions:

\begin{itemize}
  \item MurMur3 \todo{add citation/links to all hash functions}
  \item XX3
  \item XX64
  \item XX128
  \item City 1.1
  \item Farm 1.0
  \item Farm 1.1
  \item Wy 3
  \item Metro
\end{itemize}
 
While other hash function implementations were considered, the implementation of
all hash functions is given by \texttt{zero-allocation-hashing} 0.16. The
benchmark was executed in throughput mode using a fork count of 2. Besides that,
default parameters were used.


\section{Datasets used}
To evaluate properties of FracMinHash, analysis was performed using different
data sets. 

The first dataset (\textbf{A}) consists of 128 different \textit{Phytophthora}
genomes. The list is taken from \cite{mandalComparativeGenomeAnalysis2022}
without further modifications. Genomes were downloaded from NCBI uding the
\texttt{datasets} utility \cite{sayersDatabaseResourcesNational2022} using the
accession codes listed in that study.

The second dataset (\textbf{B}) consists of 72 mtDNA sequences of
\textit{Phytophthora}
and other Oomycetes of the Peronosporaceae family. The list is taken without
further modifications from Supplementary Table 1 from
\cite{winkworthComparativeAnalysesComplete2022}. Genomes were downloaded from
NCBI using the Entrez interface \cite{sayersDatabaseResourcesNational2022} using
the accession codes listed in that study, appended by the identifier of the most
recent version for that accession code.

The third dataset (\textbf{C}) consists of all 64 \textit{Phytophthora}
reference sequences in the NCBI databasase ("reference") as well as five
different query sequences that are typically found in soil samples of avocado
orchards that are infected with \textit{Phytophthora cinnamomi}
\cite{solis-garciaPhytophthoraRootRot2020}. Those query sequences are divided
into two bacterial genomes, two fungal and one \textit{Phytophthora cinnamomi}
genome:

\todo{add genome sizes}
\begin{itemize}
  \item fungal: \textit{Mortierella claussenii} - GCA\_022750515.1
  \item bacterial: \textit{Thermogemmatispora aurantia} - GCA\_008974285.1
  \item fungal: \textit{Venturia carpophila} - GCA\_014858625.1
  \item bacterial: \textit{Pseudomonas syringae} - GCA\_018394375.1
  \item fungal: \textit{Phytophthora cinnamomi} - GCA\_001314365.1
\end{itemize}

The full list of reference sequences can be found in the appendix \todo{ref}.
The reference sequences were downloaded using the web interface (Filter:
reference sequences), the query sequences were downloaded using the
\texttt{datasets} utility.



\section{Comparison with published Phylogenies}
\cite{mandalComparativeGenomeAnalysis2022} displays a rooted phylogenetic tree
in Figure 4 that was generated using the \texttt{mashtree}
\cite{katzMashtreeRapidComparison2019,ondovMashFastGenome2016}. Unfortunately,
the corresponding distance matrix or a serialized version of the tree are not
available. Thus, the data needs to be re-computed. For this, the
\texttt{mashtree\_bootstrap.pl} 1.4.6 was applied to dataset A as described by
\cite{mandalComparativeGenomeAnalysis2022}, additionally the distance matrix was
saved using the \texttt{--outmatrix} parameter.

Distance matrices for dataset A using the FracMinHash method were calculated using different
combinations of the scaling parameter $s$ and $k$-mer size $k$:

\begin{itemize}
  \item $k=10$, $s=2000$
  \item $k=19$, $s=2000$
  \item $k=20$, $s=2000$
  \item $k=21$, $s=2000$
  \item $k=25$, $s=2000$
  \item $k=30$, $s=2000$
  \item $k=21$, $s=500$
  \item $k=21$, $s=1000$
  \item $k=21$, $s=40000$
\end{itemize}

To ensure comparability with the \texttt{mashtree} results, the same hashing
function (MurMur) and random seed (42) were used to calculate the sketches
\cite{katzMashtreeRapidComparison2019,ondovMashFastGenome2016}.

For all distance matrices, SplitsTree 6.0.0
\cite{husonApplicationPhylogeneticNetworks2006} was used to obtain trees, splits
and outlines. Trees were calculated using the Neighbor Joining method
\cite{saitouNeighborjoiningMethodNew1987}. Splits were obtained using the
Neighbor Net method
\cite{bryantNeighborNetAgglomerativeMethod2004,bryantNeighborNetImprovedAlgorithms2023}
using the default parameters. Based on this, a phylogenetic outline was
calculated \cite{bagciMicrobialPhylogeneticContext2021}.
\todo{The Consensus Network method (Holland et al 2004) was used (default options, except Consensus=ConsensusNetwork) so as to obtain 300 splits, incompatible.}



\section{Split difference analysis}
\label{sec:splitanalysis}
To analyse the differences between the phylogenetic outline based on the Mash
distances and the outlines based on FracMinHash further, the splits were
analysed in details. For this, the splits were exported from SplitsTree
6.0.0-alpha \cite{husonApplicationPhylogeneticNetworks2006} using the plain text
format.

Those files were processed with the script \texttt{compare\_splity.py}. This is
based on \texttt{python} 3.12 \todo{citation} and \texttt{pandas} 2.1.4
\cite{PandasdevPandasPandas2024,mckinneyDataStructuresStatistical2010}. All sets
of splits are compared pairwise, that is for two sets $\Sigma_A$ and $\Sigma_B$,
the following properties are calculated:

\begin{itemize}
  \item the total sum of weight for all splits in $\Sigma_A$ and $\Sigma_B$,
  respectively: $\sum_{s \in \Sigma_A}{\omega(s)}$ and $\sum_{s \in
  \Sigma_B}{\omega(s)}$ 
  \item the number of splits in $\Sigma_A / \Sigma_B$ and the number of splits
  in $\Sigma_B / \Sigma_A$
  \item the total weight difference associated with the remaining splits, i.e.
  $\sum_{s \in \Sigma_A / \Sigma_B}{\omega(s)}$ and $\sum_{s \in \Sigma_B /
  \Sigma_A}{\omega(s)}$
  \item the Robinson-Foulds distance
  \cite{robinsonComparisonPhylogeneticTrees1981} of the two sets of splits, i.e.
  $D_{RF} = \frac{1}{2}|(\Sigma_A / \Sigma_B) \cup (\Sigma_B / \Sigma_A)|$
\end{itemize}

The script outputs a list of the diverging splits sorted by the weight of the
split. Each split is formatted as a list of taxa names on the smaller side of
the split sorted alphabetically and joined using the "|" symbol such that the
split can be searched in SplitsTree 6 using the regular expression search.



\section{Reproducing phylogenies for shorter sequences}
To get an idea of the practical lower boundarys of FracMinHash in terms of input
genome size, dataset B was sketched. As this dataset consists of just a single
long FASTA file, it was split into its records using \texttt{split-fasta} 1.0.0
\todo{check citation}. The files were then sketched with FracMinHash using
$k=21$, $s \in \{1, 10, 50, 100, 1000\}$ and the \todo{redo with most recent
implementation, check sketch sizes. Redo because I don't know with which version
the sketches were computed, thus I don't know how to calculate the sizes.}
MurMur hash function with random seed 42.

Given the calculated distances, SplitsTree 6.0.0-alpha
\cite{husonApplicationPhylogeneticNetworks2006} was used to obtain trees, splits
and outlines. Trees were calculated using the Neighbor Joining method
\cite{saitouNeighborjoiningMethodNew1987}. Splits were obtained using the
Neighbor Net method
\cite{bryantNeighborNetAgglomerativeMethod2004,bryantNeighborNetImprovedAlgorithms2023}
using the default parameters. Based on this, a phylogenetic outline was
calculated \cite{bagciMicrobialPhylogeneticContext2021}.


\section{Calculating distances of distantly related genomes and genomes with different sizes}
To analyse the claim that FracMinHash works better for genomes of different
sizes \todo{ensure this claim is cited at least once!}, both reference and query
sequences of dataset C were sketched, distances calculated and the results
compared. 

\subsection*{Sketching using FracMinHash}
\label{sec:intersectionlog}
Reference and query sequences were sketched with $k=21$, $s=2000$ using FarmHash
with random seeds $rs \in \{10, 20, 30, 40, 50\}$. As the default sketch size
for Mash is 10000 \cite{ondovMashFastGenome2016} and to ensure that it is not
the fact that the relevant sketches are just larger, additional sketching with
$s=500$ (which aims to bring the sketch size of the bacterial query sequences to
10000) and $s=3500$ (which aims to bring the sketch size of the fungal query
sequences to 10000) was performed. Using the sketches, the distance matrix was
calculated. For this, the implementation was changed such that an additional log
is created that lists all empty intersections when calculating the intersection
of the two sketches.

\subsection*{Sketching using MashTree}
To obtain a base value to compare agains, \texttt{mashtree} 1.4.6
\cite{katzMashtreeRapidComparison2019,ondovMashFastGenome2016} was applied with
hash seed $rs \in \{10, 20, 30, 40, 50\}$, \texttt{--outmatrix} and the
\texttt{--save-sketches} parameter. The resulting sketches were then also
converted into JSON format using \texttt{mash info -d}.

\subsection*{Visual comparison and split differences}
Given all calculated distances, SplitsTree 6.0.0-alpha
\cite{husonApplicationPhylogeneticNetworks2006} was used to obtain trees, splits
and outlines. Trees were calculated using the Neighbor Joining method
\cite{saitouNeighborjoiningMethodNew1987}. Splits were obtained using the
Neighbor Net method
\cite{bryantNeighborNetAgglomerativeMethod2004,bryantNeighborNetImprovedAlgorithms2023}
using the default parameters. Based on this, a phylogenetic outline was
calculated \cite{bagciMicrobialPhylogeneticContext2021}.

The splits were then also analysed using the script outlined in Section
\ref{sec:splitanalysis}.

\subsection*{Obtaining reference values}
As there are no published phylogenies for dataset C that enable a comparison, I
need a different ground truth to compare the results of above distance
calculation against. As the aim for this experiment is to analyze the influcence
of different genome sizes of distantly related organisms, I have limited this
analysis to the five query sequences of dataset C and the reference genome for
\textit{Phytophthora infestans} (GCF\_000142945.1) as this is one of the largest
genomes in the set of reference sequences. For those sequences, I have prepared
two different values to compare against.

The first is the Average Nucleotide Identity (ANI) calculated with
\texttt{OrthoANI} \cite{leeOrthoANIImprovedAlgorithm2016} using the default
parameters with \texttt{BLAST+} 2.14.1
\cite{camachoBLASTArchitectureApplications2009} for alignment.

The second is the Average Amino Acid Identity (AAI)
\cite{konstantinidisGenomeBasedTaxonomyProkaryotes2005} calculated with the
\texttt{Enveomics Collection} online resource
\cite{rodriguez-rEnveomicsCollectionToolbox2016} using default parameters. As
this requires amino acid sequences as input and those are not available for the
\textit{Venturia carpophila} and \textit{Phytophthora cinnamomi} sequences, they
were predicted using the \texttt{gmes\_petap.pl} script provided in the
\texttt{GeneMark ES} 4.72 package \cite{lomsadzeGeneIdentificationNovel2005}
using the default parameters and extracted using the \texttt{gffread} utility
0.12.7 \cite{perteaGFFUtilitiesGffRead2020}.


\subsection*{Calculating empty intersections}
When the sketch intersections are empty, the estiamted Jaccard similarity is
always 0. Mash does not only calculate the sketch intersections, but also
intersects that result with the sketch of the union of the two input sketches
\cite{ondovMashFastGenome2016}. This increases the chances of getting a Jaccard
estimation of 0. 

To test this, I have prepared a the script \texttt{get\_empty\_intersections.py}
which takes a list of Mash sketches in JSON format and outputs all pairs for
which the numerator of the Jaccard estimation is empty.

The same is possible using the added log output generated by the distance
calculation described in Section \ref{sec:intersectionlog}.

Using the five different random seeds, we can convert this output into a table
that lists the number of empty intersections for each pairwise calculation.
\todo{only sketches with the same random seed are compared against each other}



\section{Hash Count analysis}
\subsection*{$k$-mer Coordinates}
To analyse the distribution of $k$-mers that are part of a FracMinHash sketch,
the coordinates of each $k$-mer were exported when sketching dataset C with
$k=21$, $s=2000$ and FarmHash with random seed $rs \in \{10, 20, 30, 40, 50, 60,
70, 80, 90\}$. The larger amount of different hash seeds in comparison with the
other experiments is to reduce the impact of a single seed further. 
The coordinates saved for each $k$-mer consist of the following:

\begin{itemize}
  \item the sequence index, the index of the sequence inside the FASTA file
  \item the $k$-mer index relative to the file, $k$-mers including
  ambiguous bases are not counted, no $k$-mer is spanning two or more sequences
  in the FASTA file
  \item the $k$-mer index relative to the current sequence, $k$-mers including
  ambiguous bases are not counted
  \item the $k$-mer index relative to the file, $k$-mers including ambiguous
  bases are counted
  \item the $k$-mer index relative to the current sequence, $k$-mers including
  ambiguous bases are counted
  \item the $k$-mer itself, not the canonical form of it
  \item the hash value of the canonical $k$-mer
\end{itemize}

The coordinate files were then analyzed with the script
\texttt{coordinates\_complexity\_correlation.py} using \texttt{python} 3.12
\todo{citation} and \texttt{numpy} 1.26.3 \cite{ArrayProgrammingNumPy}. For this
analysis, each sequence in the FASTA is split into non-overlapping windows of
size $w=10000$. For each window, the number of hashes inside the window that are
part of the sketch are counted including duplicates and ambiguous bases.

\subsection*{Sequence Complexity}
Also part of the analysis is the the sequence complexity for each window. For
this, the \texttt{macle} tool \cite{pirogovHighcomplexityRegionsMammalian2019}
was used. First, the single FASTA file was split into the individual sequences
using \texttt{split-fasta} 1.0.0 \todo{check citation}. Then, the complexity was
calculated using the default parameters of \texttt{macle} and a a window size of
$w=10000$. The output identifies each window by the position of its center base,
so this needs to be shifted to the front of the window to match them to the hash
count windows. \texttt{macle} prints a complexity of $-1$ if the window contains
ambiguous bases, thus this information was used to identify and discard those
windows.

\subsection*{Statistical Analysis of connection between hash counts and sequence complexity}
For all windows with non-negative sequence complexity $C_m$, the
Pearson-Correlation \todo{check format and citation} was calculated. 

Next, to check if there is any connection between windows having an unusual hash
count and having a low sequence complexity, we need to define what we mean by
that: Given the scaling parameter $s$ and a window size $w$, we would expect on
average $\frac{w}{s}$ hashes per window that are part of the sketch. A window
has \textbf{unsual hash counts} if the number of hashes is outside the interval
$[\frac{w}{s} - r \frac{w}{s}, \frac{w}{s} + r \frac{w}{s}]$. I have set
$r=0.95$, $s=2000$ and $w=10000$, which classifies all windows with a hash count
of $0$ or $\geq 10$ as unusual.

Given those inputs, I have performed the Mann-Whitney-U Test
\cite{mannTestWhetherOne1947,wilcoxonIndividualComparisonsRanking1945} with the
complexity values of each window into the two categories \textbf{unusual count}
and \textbf{usual count}. I have corrected $\alpha = \frac{0.05}{69}$ to account
for multiple testing of the 69 different genomes.

The statistics were computed using \texttt{scipy} 1.11.3
\cite{virtanenSciPyFundamentalAlgorithms2020}.

\subsection*{Analysis of potential connection between low hash counts and gene location}
To check if unusual hash counts can be linked to the location of (effector)
genes in \textit{Phytophthora infestans}, I have mapped the start position of
each CDS annotation to the corresponding window and counted the number of CDS
per window.

\section{Runtime benchmarks}
To compare the runtime of \texttt{fmhdist} with \texttt{mash} and sourmash,
\todo{add details}